Add docs for configuring CloudKitty (#1036)
stackhpc/stackhpc-kayobe-config: doc/source/operations/cloudkitty.rst

==========
CloudKitty
==========

Configuring in kayobe-config
============================

> By default, CloudKitty uses Gnocci and Ceilometer as the collector and fetcher

Reviewer: ```suggestion
By default, CloudKitty uses Gnocchi and Ceilometer as the collector and fetcher
```

backends. Unless the system has a specific reason not to, we recommend instead
using Prometheus as the backend for both. The following instructions explain

> how to do this

Reviewer 2: Nit-Pick: Needs either a full stop or a colon.

```suggestion
how to do this.
```

Enable CloudKitty and disable InfluxDB, as we are using OpenSearch as the
storage backend. Set the following in ``kolla.yml``:

> .. code-block:: console

Reviewer 3: ```suggestion
.. code-block:: yaml
```

  kolla_enable_cloudkitty: true

>   # Explicitly disable influxdb as we are using OpenSearch as the CloudKitty backend

Reviewer 2: Not sure if this is necessary, as long as it isn't made the backend then it should be fine, no? 
Leaving it commented out may be sufficient?

Author: Sure it will be fine in the sense that nothing will break. But when KA sees that CloudKitty is enabled, it will also deploy InfluxDB by default, so we'd have an InfluxDB container just sat there doing nothing.

Author: Really it would be good to patch KA to fix this properly, but I recall there being some complication that made this non-trivial

>   kolla_enable_influxdb: false

Reviewer 3: Maybe we should change this default? The default in k-a is a bit more sensible but kayobe doesn't have all the info:

```
enable_influxdb: "{{ enable_cloudkitty | bool and cloudkitty_storage_backend == 'influxdb' }}"
```

Author: Will leave these docs as-is for now, we've agreed to look into improving CloudKitty-related defaults in a future PR.

> Use Prometheus as the collector and fetcher backends, and Elasticsearch as the

Reviewer 2: Nit-Pick: Might sound better like this?
```suggestion
Set Prometheus as the backend for both the collector and fetcher, and Elasticsearch as the
```

> storage backend. Note that this is patched in our fork of CloudKitty to also

Reviewer 2: Don't we exclusively use `OpenSearch`, as well as we aren't able to set `cloudkitty_storage_backend: opensearch` as the backend, as it won’t work? This is to say that this sentence suggests that "opensearch" can be set for `cloudkitty_storage_backend:` when, in fact, nothing needs changing if you want to use `opensearch`.

Maybe something like: In our fork of CloudKitty `cloudkitty_storage_backend: elasticsearch` is patched to correspond to both ElasticSearch as well as OpenSearch.

Author: Discussed this in more detail with Max in person, will update this section to make things clearer.

work with OpenSearch. Proper support is still pending in Kolla-Ansible `here
<https://review.opendev.org/c/openstack/kolla-ansible/+/898555>`__. If you have

> TLS enabled, you will also need to allow insecure connections for Prometheus

Reviewer 3: No no no :)

```
cloudkitty_prometheus_cafile: "{{ openstack_cacert }}"
cloudkitty_elasticsearch_cafile: "{{ openstack_cacert }}"
```

Even this shouldn't be necessary - it should be the default like it is for other services.

Author: I'd tried this before and was still seeing errors, afraid I didn't write them down. I'll see if I can recreate in a test env, as you say it seems like the defaults should work

Reviewer 3: We have it working in at least one deployment

Reviewer 3: Although that was explicitly setting the options in cloudkitty.conf rather than using those variables.

Author: I think I've found the issue, Kayobe did not have a value set for ``openstack_cacert`` when I tried to deploy because ``OS_CACERT`` was not set. I'm going to update the docs here to state that ``openstack_cacert`` should be set in Kayobe globals too.
On this point, I do wonder if we should alter defaults in SKC to handle this better.

Reviewer 3: Slightly confusingly, `openstack_cacert` doesn't set `openstack_cacert` in kolla-ansible. Typically we set `openstack_cacert` in kolla-ansible as part of the internal TLS deployment.

Author: Confusing, but explains my problems. Thanks

and Elasticsearch. Set the following in ``kolla/globals.yml``:

> .. code-block:: console

Reviewer 3: ```suggestion
.. code-block:: yaml
```

> Set Prometheus as the backend for both the collector and fetcher, and Elasticsearch as the
> storage backend. Note that this is patched in our fork of CloudKitty to also
> work with OpenSearch. Proper support is still pending in Kolla-Ansible `here
> <https://review.opendev.org/c/openstack/kolla-ansible/+/898555>`__. If you have
> TLS enabled, you will also need to allow insecure connections for Prometheus
> and Elasticsearch. Set the following in ``kolla/globals.yml``:
> 
> .. code-block:: console
> 
>   cloudkitty_collector_backend: prometheus
>   cloudkitty_fetcher_backend: prometheus
>   cloudkitty_storage_backend: elasticsearch
>   cloudkitty_elasticsearch_insecure_connections: true
>   cloudkitty_prometheus_insecure_connections: true

Reviewer 2: ```suggestion
Set Prometheus as the backend for both the collector and fetcher, and Elasticsearch as the
storage backend; this is achieved by configuring the following within ``kolla/globals.yml``:

.. code-block:: console

  cloudkitty_collector_backend: prometheus
  cloudkitty_fetcher_backend: prometheus
  cloudkitty_storage_backend: elasticsearch

Note that in our fork of CloudKitty ``cloudkitty_storage_backend: elasticsearch`` is patched to correspond to both ElasticSearch as well as OpenSearch
without having to change the ``cloudkitty_storage_backend:`` variable. 
Proper support of this is still pending for Kolla-Ansible `here
<https://review.opendev.org/c/openstack/kolla-ansible/+/898555>`__. 
Furthermore, if you have TLS enabled, you will also need to allow insecure connections for Prometheus and Elasticsearch. 

.. code-block:: console

  cloudkitty_elasticsearch_insecure_connections: true
  cloudkitty_prometheus_insecure_connections: true
```

Author: Seems sensible to split these. Will make the change.

The default collection period is one hour, which is likely too long for your
system. CloudKitty will charge for the entire period if any usage is seen
within this timeframe. This means that even just one minute will be counted as
an hour's usage. It is recommended to change this to a lower number, such as
ten minutes. When using Prometheus as the collector, you need to change the
scope_key to match the metrics provided by the Prometheus OpenStack Exporter.

> Set the following in ``cloudkitty.conf``:

Reviewer 2: ```suggestion
The default collection period is one hour, which is likely too long for most systems
as CloudKitty charges by the **entire** collection period if any usage is seen within
this timeframe. This is regardless of actual usage, meaning that even one minute will
be charged as an full hour's usage. As a result, it is recommended to adjust the
collection interval, ``period`` (in units of seconds), appropriatelly (e.g. ten minutes).
Furthermore, when using Prometheus as the collector, you need to change the
``scope_key`` to match the metrics provided by the Prometheus OpenStack Exporter.
Both of these can be achieved by setting the following in
``kolla/config/cloudkitty.conf``:
```

Author: Looks good, have made these changes

> Set the following in ``cloudkitty.conf``:

Reviewer 3: ```suggestion
Set the following in ``kolla/config/cloudkitty.conf``:
```

.. code-block:: console

  [collect]
  scope_key = tenant_id
  period = 600

> You will need to configure which metrics CloudKitty should track. This example
> will track for flavors and volumes, set in
> ``kolla/config/cloudkitty/metrics.yml``:

Reviewer 2: ```suggestion
You will need to configure which metrics CloudKitty should track. The following
example, set in ``kolla/config/cloudkitty/metrics.yml``, will track for VM flavors and
the total utilised volume.
```

> .. code-block:: console

Reviewer 3: ```suggestion
.. code-block:: yaml
```

>   metrics:

Reviewer 3: Indentation looks a bit inconsistent & broken here (e.g. mutate_map)

Author: Fixed

    openstack_nova_server_status:
        alt_name: instance
        groupby:
        - uuid
        - user_id
        - tenant_id
        metadata:
        - flavor_id
        - name
        mutate: MAP
        mutate_map:
        0.0: 1.0  # ACTIVE
        11.0: 1.0 # SHUTOFF
        12.0: 1.0 # SUSPENDED
        16.0: 1.0 # PAUSED
        unit: instance
    openstack_cinder_limits_volume_used_gb:
        alt_name: storage
        unit: GiB
        groupby:
        - tenant_id

TODO: Should we explain how to handle OpenStack exporter metrics relabelling,
for the case where a system may have had Monasca deployed in the past?

Post-configuration with openstack-config
========================================

This is an example `openstack-config
<https://github.com/stackhpc/openstack-config>`__ setup to create mappings for

> the metrics configured above. Note that the costs are scaled for the ten-minute

Reviewer 2: ```suggestion
the metrics configured above. Note that the costs are scaled for the ten minute
```

collection period, e.g. a flavor with 1 VCPU will cost 1 unit per hour.

.. code-block:: yaml

  # Map flavors based on VCPUs
  openstack_ratings_hashmap_field_mappings:
    - service: instance
        name: flavor_id
        mappings:
        - value: '1' # tiny compute flavor (1 vcpu) with an OpenStack flavor ID of 1

>             cost: 0.1666666666666666

Reviewer 4: :nauseated_face:

Author: Perhaps jinja can template this, don't think it's really an issue though.
